docs/topics/complex_plugins.rst

```diff
@@ -0,0 +1,79 @@
+..  _complex_plugins:
+
+
+##########################
+Complex plugin integration
+##########################
+
+****************************************
+Overview of how django CMS wraps plugins
+****************************************
+
+A django CMS plugin editing interface is fundamentally a Django admin interface for the model object in question,
+wrapped with some additional layers to present it in a way that's appropriate to the editing task and workflow.
+
+The wrapping looks a bit like this:
+
+* ``<div class="cms-modal-iframe>...</div>``, a modal dialog box, containing
+
+  * ``<iframe...>``, which contains
+
+    * ``<html>...</html>``, the actual Django admin page
+
+In order to present a consistent interface to the user, django CMS overrides some parts of the interface Django
+provides.
+
+For example, the Django admin provides a ``<div class="submit-row">`` for forms, that contains **Save** and **Delete**
+buttons. In a standard Django admin page, **Save** is actually an ``<input type="submit">`` and **Delete** is simply a
+link.
+
+django CMS hides the ``div.submit-row`` (with ``visibility: none``) and instead provides its own buttons: **Cancel**,
+**Delete** and **Save**.
+
+The plugin wrapper moves the functionality of Django buttons is moved from within the ``<html>...</html>`` to the
+``<div class="cms-modal-iframe>...</div>``.
+
+As far as the user is concerned, the wrapper must (and must) behave transparently. For example, **Save** must do what
+it would in a standard admin form page, including validation and cleaning, and the same for any inlined admin objects.
+
+This allows django CMS to perform some additional actions, such as tidying up the structure mode display when the
+modal  dialog box is closed.
+
+How wrapping is implemented
+***************************
+
+In order to achieve this, django CMS refers the **Save** and **Delete** actions to the admin form.
+
+Hitting django CMS's **Save** button invokes the ``<input>`` on the admin form, as though it had been invoked in the
+browser by the user, by assigning its own **Save** button to the form's ``submit`` event.
+
+.. todo: Anything else we should note?
```

Comment:
probably shouldn't, but here are few more things we do:

the cms does some hidden things like this for example https://github.com/divio/django-cms/blob/release/3.4.x/cms/static/cms/js/modules/cms.modal.js#L820-L849

or for example there are certain conditions for the modal to be closed after "saving" that may seem non-obvious https://github.com/divio/django-cms/blob/release/3.4.x/cms/static/cms/js/modules/cms.modal.js#L972-L979

or that we inject ctrl+enter shortcut to submit the modal from the keyboard even if you're not inside the form field (cmd+enter for mac)